### PR TITLE
[DEV-510] - multi-voucher issuance

### PR DESCRIPTION
**Pull request by grp06**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] I have updated the documentation accordingly. (if applicable)
- [ ] All new and existing tests passed.

### CODE REVIEW

File: src/Badges.sol [hunk 1]
```diff
@@ -291,23 +321,46 @@ contract Badges is
...
+    // this authenticates the signature coming from the issuer
+    bytes32 hash = getMerkleAgreementHash(_from, _uri, _root);
+    require(
+      SignatureCheckerUpgradeable.isValidSignatureNow(_from, hash, _signature),
+      "safeCheckMerkleAgreement: invalid signature"
+    );
+    bytes32 leaf = keccak256(abi.encodePacked(_to));
+    // this authenticates that the claimant is indeed part of the tree whose root was signed
+    require(MerkleProof.verify(_proof, _root, leaf), "safeCheckMerkleAgreement: invalid leaf");
+    uint256 voucherHashId = uint256(keccak256(abi.encode(_from, _to)));
```
Comment (rahulrumalla) [inline src/Badges.sol:343]: We're not using 2 different ways to generate a voucherHashId... 

1. would there be collissions you think?
2. if you get a badge via multi-issuance, would the protocol prevent the user from having a badge via the single-issuance?

File: src/Badges.sol [hunk 2]
```diff
@@ -291,23 +321,46 @@ contract Badges is
...
+    return voucherHashId;
+  }
+
   function safeCheckAgreement(
     address _active,
     address _passive,
     string calldata _uri,
     bytes calldata _signature
-  ) internal virtual returns (uint256) {
+  ) internal view virtual returns (uint256) {
```
Comment (rahulrumalla) [inline src/Badges.sol:352]: Should we make this public?

File: test/badges.test.ts [hunk 3]
```diff
@@ -214,9 +220,194 @@ const deployContractFixture = async () => {
...
+    claimant,
+    randomSigner,
+    typedData,
+    specDataHolderProxy,
+    merkleTypedData,
+  }
 }
 
+describe('Merkle minting', () => {
+  it('Happy path: should allow minting when an address is whitelisted on a merkle tree', async () => {
```
Comment (rahulrumalla) [inline test/badges.test.ts:258]:
```suggestion
  it('Should allow minting when an address is allowlisted on a merkle tree', async () => {
```

Nitpick

File: src/Badges.sol (hunk 1 above)
Comment (grp06) [reply src/Badges.sol:343]: 1. I don't think there would be collisions... but to maximize probability of uniqueness, we might as well just throw in the `MERKLE_AGREEMENT_HASH` and `uri`... wouldn't hurt. 
2. with that, we definitely won't have collisions or issues with single/multi-issuance conflicting

File: src/Badges.sol (hunk 2 above)
Comment (rahulrumalla) [reply src/Badges.sol]: Hmm.. yeah. I think this then an internal operation. The safeCheckMerkleAgreement also should be internal I blv. Meaning the we need to test it with a different setup. Could you look into that, please?

File: src/Badges.sol [hunk 4]
```diff
@@ -424,7 +422,7 @@ contract Badges is
     );
 
     // this authenticates that the claimant (leaf) is indeed part of the tree whose root was signed
-    bytes32 leaf = keccak256(abi.encodePacked(_to)); //todo:: why not use abi.encode? which is appropriate?
+    bytes32 leaf = keccak256(abi.encodePacked(_to));
```
Comment (grp06) [inline src/Badges.sol:425]: @rahulrumalla I looked into this. From what I read, as long as you're only encoding one argument, this is the way to go, uses slightly less gas.

**Review by rahulrumalla — APPROVED** [review]

**Pull request merged**